jax/_src/typing/ndarray.py

```diff
@@ -292,3 +295,14 @@ def weak_type(self) -> bool: ...
 for t in device_array.device_array_types:
   ndarray.register(t)
 ndarray.register(pxla._SDA_BASE_CLASS)
+
+
+class HasArrayMethod(Protocol):
+  def __array__(self) -> np.ndarray:
+    ...
+
+
+ArrayLike = Union[bool, int, float, complex, ndarray, core.Tracer, np.ndarray, HasArrayMethod]
```

Comment:
`SupportsArray` might be better than `HasArrayMethod` -- it aligns with stdlib https://docs.python.org/3/library/typing.html#protocols and numpy https://github.com/numpy/numpy/blob/6094eff95c19c433648593f3cb6bca489396f9ae/numpy/_typing/_array_like.py#L36-L38.
BTW, we can add a comment about why not directly use `Union[Array, numpy.typing.ArrayLike]` -- JAX ArrayLike does NOT allow (nested) sequences, `str` or `bytes`.

Reply:
None of this is public API yet aside from `jax.Array`, so there's no issue with adjusting names if we wish.

We don't use `numpy.typing.ArrayLike` here because it's too permissive (e.g. it allows lists).

Reply:
Yes, JAX does not allow nested sequence. I just suggest that we can add a comment in code about this reason(and/or a link to docs).

Reply:
Thanks for the suggestions - changes in https://github.com/google/jax/pull/12813